core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -721,6 +723,15 @@ class SparkContext(config: SparkConf) extends Logging {
   } catch {
     case NonFatal(e) =>
       logError("Error initializing SparkContext.", e)
+      if(_env == null && envCreated != null) {
```

Comment:
How can this happen? I think the issue has to be understood thoroughly

Reply:
@HyukjinKwon I looked through the SparkContext creation again and here are my thoughts. There is a window where `SparkEnv` can be partially initialized (thus starting the RPC server) yet never assigned to `_env` in time for the usual shutdown logic to call `_env.stop()`. 
By stopping `envCreated` directly if `_env` remains null on exception, we can ensure that the RPC server (and other env resources) get shut down. That prevents the leftover port-blocking issue.
Please correct me if I'm wrong.

Reply:
Agree with @HyukjinKwon.
The code, as formulated, does not do what you are describing @soumasish.
`envCreated` is assigned *after* `createSparkEnv` returns - and immediate is assigned to `_env`.
So `envCreated` is always same as `_env`.

If the concern is initialization of individual components of SparkEnv can fail (which can happen), we will need to handle it in `SparkEnv.create`.
Surround it in try/finally - and stop each component which was successfully initialized when the method is going to throw exception.